_source/_posts/2019-12-20-five-tools-improve-java.adoc

```diff
@@ -0,0 +1,345 @@
+                <type>jar</type>
+            </dependency>
+        </oldVersion>
+        <parameter>
+            <onlyModified>true</onlyModified>
+            <!--2-->
+            <breakBuildOnBinaryIncompatibleModifications>true</breakBuildOnBinaryIncompatibleModifications>
+            <!--3-->
+            <breakBuildBasedOnSemanticVersioning>true</breakBuildBasedOnSemanticVersioning>
+            <!--4-->
+            <postAnalysisScript>src/japicmp/postAnalysisScript.groovy</postAnalysisScript>
+        </parameter>
+    </configuration>
+    <executions>
+        <execution>
+            <id>japicmp</id>
+            <goals>
+                <goal>cmp</goal>
+            </goals>
+        </execution>
+    </executions>
+</plugin>
+----
+<1> The dependency block the previous version.
+<2> `breakBuildOnBinaryIncompatibleModifications` - fail the build on any backward-incompatible changes.
+<3> `breakBuildBasedOnSemanticVersioning` - fail based on semver rules. For example, if your public API changes in a way that would require a "minor" version change.
+<4> `postAnalysisScript` - Optional, allows use of a custom Groovy script to modify the results based on your own needs. If you want to allow new default methods in interfaces, you would need a script similar to this:
+
+.postAnalysisScript.groovy
+[source,groovy]
+----
+import static japicmp.model.JApiCompatibilityChange.*
+import static japicmp.model.JApiChangeStatus.*
+
+def it = jApiClasses.iterator()
+while (it.hasNext()) {
+    def jApiClass = it.next()
+
+    if (jApiClass.getChangeStatus() != UNCHANGED) {
+        def methodIt = jApiClass.getMethods().iterator()
+        while (methodIt.hasNext()) {
+            def method = methodIt.next()
+            def methodChanges = method.getCompatibilityChanges()
+            methodChanges.remove(METHOD_NEW_DEFAULT)
+        }
+    }
+}
+return jApiClasses
+----
+
+Take a look at the https://siom79.github.io/japicmp/[japicmp project documentation] for more examples.
+
+== 5. Don't Skip Code Reviews
+
+Using the above tools can help make your code reviews more effective. The goal should be to automate as much as possible out of your code review so that the human element can shine through. Have you ever asked, "is there a test for this" in a code review? Automate that, send your coverage data to a tool like https://codecov.io/[Codecov], which can add the coverage delta to your pull requests. If your project has strict code style guidelines, you can use https://checkstyle.org/[Checkstyle].
+
+Code reviews are great; they provide an opportunity for both the author and the reviewer to learn from each other and ask questions, suggest alternatives, or discuss other architectural topics. The reviewer shouldn't be wasting time checking for things that a program can detect.
+
+TIP: **Pro Tip:** Keep the number of changes in your reviews small and to the point. Nobody wants to review hundreds of potentially unrelated code changes at once.
```

Comment:
I typically let the admonitions do the work, rather than adding Note or Pro Tip in front of them. If you like adding in a prefix, more power to you!